transformer_engine/pytorch/module.py

```diff
@@ -572,6 +608,23 @@ def grad_output_preprocess(
 
         return grad_output_mat, grad_output_c, grad_output_t, grad_bias
 
+    def noop_cat(self, buffer_name: str, *params_split: Tuple[torch.Tensor, ...]) -> torch.Tensor:
+        """No-op replacement of `torch.cat`. The buffer and split parameters must occupy
+           the same memory region. If this is not the case, then the split parameters
+           are concatenated and the buffer is overwritten.
+        """
+
+        assert hasattr(self, buffer_name), f"No buffer named {buffer_name}"
+        full_param_buffer = getattr(self, buffer_name)
+        split_size = full_param_buffer.shape[0] // len(params_split)
+        for i, p in enumerate(params_split):
+            if p.data.data_ptr() != full_param_buffer[i*split_size : (i+1)*split_size].data_ptr():
+                with torch.no_grad():
+                    setattr(self, buffer_name, torch.cat(params_split))
+                break
```

Comment:
So just to be clear - this does not then "fix" the problem of the split params not being in a single buffer -> it will just perform the concat every time in that case, right?

Is it possible to update the params after this cat so that they point to the right place and on subsequent iteration concat is not needed anymore?

Reply:
Good catch, forgot to update the params memory :)

Reply:
/te-ci